### PR TITLE
Add volumeattach/volumedetach API

### PR DESCRIPTION
**Pull request by zedi-pramodh**
Longhorn volumes are attached to a node where the app is running. Since the volumes we create in eve are RWO, ie only one node can attach at a time, during failover we need to make sure the node that failed is detached from the volume.

The node that is taking over will automatically attach to the volume if no other node is attached. Also, generally its very safe to detach and then attach to avoid any potential corruption.

### CODE REVIEW

**Comment by eriknordmark** [comment]
I assume there will be a follow-up PR to call this APIs. Is that correct?

@andrewd-zededa can you review?

File: pkg/pillar/kubeapi/vitoapiserver.go [hunk 1]
```diff
@@ -313,6 +313,89 @@ func RolloutDiskToPVC(ctx context.Context, log *base.LogObject, exists bool,
 	return nil
 }
 
+// GetPVFromPVC : Returns volume name (PV) from the PVC name
+func GetPVFromPVC(pvcName string, log *base.LogObject) (string, error) {
+	// Get the Kubernetes clientset
+	clientset, err := GetClientSet()
+	if err != nil {
+		err = fmt.Errorf("failed to get clientset: %v", err)
```
Comment (rene) [inline pkg/pillar/kubeapi/vitoapiserver.go:321]: A new error it's being created here, why you just don't propagate the original error (from the `GetClientSet()` call)?

File: pkg/pillar/kubeapi/vitoapiserver.go [hunk 2]
```diff
@@ -313,6 +313,89 @@ func RolloutDiskToPVC(ctx context.Context, log *base.LogObject, exists bool,
...
+		log.Error(err)
+		return "", err
+	}
+	// Get the PersistentVolumeClaim (PVC)
+	pvc, err := clientset.CoreV1().PersistentVolumeClaims(EVEKubeNameSpace).Get(context.TODO(), pvcName, metav1.GetOptions{})
+	if err != nil {
+		log.Errorf("Error fetching PersistentVolumeClaim %s: %v", pvcName, err)
```
Comment (rene) [inline pkg/pillar/kubeapi/vitoapiserver.go:328]: You could use the same approach from line [321](https://github.com/lf-edge/eve/pull/4359/files#diff-01e7f32d4e5ab78e4a798e525ecb0ba7529ba505aebd89c3cdbea6fb32ebf333R321) and return the same logged error.

Comment (zedi-pramodh) [reply pkg/pillar/kubeapi/vitoapiserver.go:328]: Ok I will make those changes across all 3 API.

Comment (zedi-pramodh) [reply pkg/pillar/kubeapi/vitoapiserver.go]: Amended the commit

**Comment by zedi-pramodh** [comment]
Can someone bless this one, if there are no further concerns.

**Comment by OhmSpectator** [comment]
> LGTM mostly, I think we need a some comments around the volumeattachment logging to ensure they are not downgraded to Function/Trace in the future.

@andrewd-zededa, @zedi-pramodh was this comment addressed?

**Review by eriknordmark — APPROVED** [review]
LGTM

**Comment by zedi-pramodh** [comment]
> > LGTM mostly, I think we need a some comments around the volumeattachment logging to ensure they are not downgraded to Function/Trace in the future.
> 
> @andrewd-zededa, @zedi-pramodh was this comment addressed?

I don't understand what Andrew meant actually :)

File: pkg/pillar/kubeapi/vitoapiserver.go [hunk 3]
```diff
@@ -313,6 +313,92 @@ func RolloutDiskToPVC(ctx context.Context, log *base.LogObject, exists bool,
...
+func GetPVFromPVC(pvcName string, log *base.LogObject) (string, error) {
+	// Get the Kubernetes clientset
+	clientset, err := GetClientSet()
+	if err != nil {
+		err = fmt.Errorf("failed to get clientset: %v", err)
+		log.Error(err)
+		return "", err
+	}
+	// Get the PersistentVolumeClaim (PVC)
+	pvc, err := clientset.CoreV1().PersistentVolumeClaims(EVEKubeNameSpace).Get(context.TODO(), pvcName, metav1.GetOptions{})
```
Comment (OhmSpectator) [inline pkg/pillar/kubeapi/vitoapiserver.go:326]: I'm just curious what is the `context.TODO()` in this case and will it be changed later...

Comment (OhmSpectator) [reply pkg/pillar/kubeapi/vitoapiserver.go:326]: Also, I'm a little bit worried about the amount of dereferencing in one line. Could be hard to debug if one of them fails.

File: pkg/pillar/kubeapi/vitoapiserver.go [hunk 4]
```diff
@@ -313,6 +313,92 @@ func RolloutDiskToPVC(ctx context.Context, log *base.LogObject, exists bool,
...
+func GetVolumeAttachmentFromPV(volName string, log *base.LogObject) (string, string, error) {
+	// Get the Kubernetes clientset
+	clientset, err := GetClientSet()
+	if err != nil {
+		err = fmt.Errorf("failed to get clientset: %v", err)
+		log.Error(err)
+		return "", "", err
+	}
+	// List all VolumeAttachments and find the one corresponding to the PV
+	volumeAttachments, err := clientset.StorageV1().VolumeAttachments().List(context.TODO(), metav1.ListOptions{})
```
Comment (OhmSpectator) [inline pkg/pillar/kubeapi/vitoapiserver.go:354]: Is it a synchronous call to the API? Are we sure we will not have issues with the watchdog in this case if the requests are handled longer than expected? This is a generic question for all API calls here.

**Review by OhmSpectator — APPROVED** [review]
I believe @andrewd-zededa it's OK, as I cannot review it properly myself. Hence, I am approving. Nevertheless, I left several questions to better understand the code.

**Comment by eriknordmark** [comment]
This needs to be rebased on master - the SPDX check workflow has issues with misdetecting issues.

**Comment by zedi-pramodh** [comment]
> This needs to be rebased on master - the SPDX check workflow has issues with misdetecting issues.

Rebased to master and amended the commit

File: pkg/pillar/kubeapi/vitoapiserver.go (hunk 4 above)
Comment (zedi-pramodh) [reply pkg/pillar/kubeapi/vitoapiserver.go:354]: Its is synchronous call, but I do not anticipate it takes even a second to get the result. Number of volumes we have in our deployments are in single digits.

File: pkg/pillar/kubeapi/vitoapiserver.go (hunk 3 above)
Comment (zedi-pramodh) [reply pkg/pillar/kubeapi/vitoapiserver.go:326]: The err should contain the actual error if something fails.

File: pkg/pillar/kubeapi/vitoapiserver.go (hunk 4 above)
Comment (OhmSpectator) [reply pkg/pillar/kubeapi/vitoapiserver.go:354]: Ok, and what about the other API calls in the file? Like `.Delete()`... So we are pretty sure all of them run fast enough, yep?

Comment (zedi-pramodh) [reply pkg/pillar/kubeapi/vitoapiserver.go:354]: Also Kubernetes API calls execution are asynchronous by default. So only sync call is to issue the command and return back after that any action will happen in background.

Comment (zedi-pramodh) [reply pkg/pillar/kubeapi/vitoapiserver.go:354]: Any more questions ? Or can this be merged ?

**Review by eriknordmark — APPROVED** [review]
Re-run tests

**Comment by zedi-pramodh** [comment]
> The same Eden issue =((((

Those should be completely unrelated.  This code path is not even executed yet.

**Pull request merged**